rowexec: incomplete sampler cleanup on node shutdown

ccl/backupccl.TestBackupRestoreInsideMultiPodTenant [failed](https://tanzanite.cluster.engflow.com/invocations/default/9768b170-f06a-4dde-9695-83372a4a1573?testReportRun=3&testReportShard=42&testReportAttempt=1#targets-Ly9wa2cvY2NsL2JhY2t1cGNjbDpiYWNrdXBjY2xfdGVzdA==) on master @ [3d2fee67f922787a45f49a2903d086d472f5b577](https://github.com/cockroachdb/cockroach/commits/3d2fee67f922787a45f49a2903d086d472f5b577):

Fatal error:

```
panic: found 3 short-living non-stopped monitors in flow b519b343
flow b519b343 0 B (10 KiB / 0 B)
    sample-aggregator-mem 0 B
    sampler-mem 0 B
    fetcher-mem 0 B
 [recovered]
	panic: found 3 short-living non-stopped monitors in flow b519b343
flow b519b343 0 B (10 KiB / 0 B)
    sample-aggregator-mem 0 B
    sampler-mem 0 B
    fetcher-mem 0 B


```
Stack: 

```
goroutine 31565341 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0052c4db0?, {0x81b0698, 0xc034e5a4e0})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x65
panic({0x622f6a0?, 0xc03499c580?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc01ba89a40, {0x81b0698, 0xc0350ee0f0}, 0x1)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657 +0x653
github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup(0xc01f1a8288, {0x81b0698, 0xc0350ee0f0})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:717 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup(0xc035059ad0, {0x81b0698, 0xc0350ee0f0})
	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499 +0xfe
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RemoteFlowRunner).RunFlow(0x81b0660?, {0x81b0698, 0xc0350ee0f0}, {0x822f1e0, 0xc035059ad0})
	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/remote_flow_runner.go:143 +0xeb
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow.func1(0xc01c4fe008, 0xc0052c53e8, 0xc01cd18408, 0xc0173036c0, {0x81b0698, 0xc034e5a660})
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:666 +0x2b0
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow(0xc01c4fe008, {0x81b0698, 0xc0350ee0f0}, 0xc0173036c0)
	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:667 +0x207
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1({0x81b0698?, 0xc034e5a660?}, {0x65ad7e0?, 0xc0173036c0?})
	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:597 +0xcb
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12({0x81b0698, 0xc034e5a660}, {0x65ad7e0, 0xc0173036c0}, 0xc0108ad880, 0xc0344ccc60)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97 +0x4b2
google.golang.org/grpc.getChainUnaryHandler.func1({0x81b0698, 0xc034e5a660}, {0x65ad7e0, 0xc0173036c0})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3({0x81b0698, 0xc034e5a660}, {0x65ad7e0, 0xc0173036c0}, 0xc0108ad880?, 0xc0342a4d80)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169 +0x76
google.golang.org/grpc.getChainUnaryHandler.func1({0x81b0698, 0xc034e5a660}, {0x65ad7e0, 0xc0173036c0})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor({0xc005e64000?, {{0x4703a5?}, {0x81fdc50?, 0xc6d8a80?}}}, {0x81b0698, 0xc034e5a4e0}, {0x65ad7e0, 0xc0173036c0}, 0xc0108ad880, 0xc0342a4d40)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95 +0x1e4
google.golang.org/grpc.getChainUnaryHandler.func1({0x81b0698, 0xc034e5a4e0}, {0x65ad7e0, 0xc0173036c0})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1({0x81b0698?, 0xc034e5a4e0?})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136 +0x36
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc003831950, {0x81b0698, 0xc034e5a4e0}, {0xc0342a4cc0?, 0xc012c35968?}, 0xc012c35960)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337 +0xcb
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1({0x81b0698?, 0xc034e5a4e0?}, {0x65ad7e0?, 0xc0173036c0?}, 0xc0108ad880?, 0xc0344ccc60?)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134 +0x8f
google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1({0x81b0698, 0xc034e5a4e0}, {0x65ad7e0, 0xc0173036c0}, 0xc0108ad880, 0x70?)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1145 +0x85
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler({0x66852c0, 0xc01c4fe008}, {0x81b0698, 0xc034e5a4e0}, 0xc00ec1ed90, 0xc00b1333a0)
	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:599 +0x143
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00f2a32c0, {0x8208860, 0xc0030131e0}, 0xc033c68a20, 0xc01553a510, 0xbf37cc0, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1335 +0xdb8
google.golang.org/grpc.(*Server).handleStream(0xc00f2a32c0, {0x8208860, 0xc0030131e0}, 0xc033c68a20, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1712 +0x9da
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:947 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 17724741
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:958 +0x136
```
<details><summary>Log preceding fatal error</summary>
<p>

```
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:598
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Cleanup
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:717
*   | github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Cleanup
*   | 	github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:499
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RemoteFlowRunner).RunFlow
*   | 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/remote_flow_runner.go:143
*   | github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:666
*   | github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow
*   | 	github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:667
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:597
*   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.ServerInterceptor.func12
*   | 	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:97
*   | google.golang.org/grpc.getChainUnaryHandler.func1
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154
*   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
*   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:169
*   | google.golang.org/grpc.getChainUnaryHandler.func1
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154
*   | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
*   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:95
*   | google.golang.org/grpc.getChainUnaryHandler.func1
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1154
*   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
*   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:136
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:337
*   | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
*   | 	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:134
*   | google.golang.org/grpc.NewServer.chainUnaryServerInterceptors.chainUnaryInterceptors.func1
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1145
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler
*   | 	github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/bazel-out/k8-fastbuild/bin/pkg/sql/execinfrapb/execinfrapb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:599
*   | google.golang.org/grpc.(*Server).processUnaryRPC
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1335
*   | google.golang.org/grpc.(*Server).handleStream
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1712
*   | google.golang.org/grpc.(*Server).serveStreams.func1.1
*   | 	google.golang.org/grpc/external/org_golang_google_grpc/server.go:947
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) found 3 short-living non-stopped monitors in flow b519b343
*   | flow b519b343 0 B (10 KiB / 0 B)
*   |     sample-aggregator-mem 0 B
*   |     sampler-mem 0 B
*   |     fetcher-mem 0 B
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=42</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreInsideMultiPodTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40101

--- a/pkg/sql/backfill/backfill.go
+++ b/pkg/sql/backfill/backfill.go
@@ -696,7 +696,8 @@ func (ib *IndexBackfiller) InitForDistributedUse(
 	return ib.init(evalCtx, predicates, colExprs, mon)
 }
 
-// Close releases the resources used by the IndexBackfiller.
+// Close releases the resources used by the IndexBackfiller. It can be called
+// multiple times.
 func (ib *IndexBackfiller) Close(ctx context.Context) {
 	if ib.mon != nil {
 		func() {
@@ -705,6 +706,7 @@ func (ib *IndexBackfiller) Close(ctx context.Context) {
 			ib.muBoundAccount.boundAccount.Close(ctx)
 		}()
 		ib.mon.Stop(ctx)
+		ib.mon = nil
 	}
 }
 

--- a/pkg/sql/backfill/mvcc_index_merger.go
+++ b/pkg/sql/backfill/mvcc_index_merger.go
@@ -503,6 +503,9 @@ func (ibm *IndexBackfillMerger) Resume(output execinfra.RowReceiver) {
 	panic("not implemented")
 }
 
+// Close is part of the execinfra.Processor interface.
+func (*IndexBackfillMerger) Close(context.Context) {}
+
 // NewIndexBackfillMerger creates a new IndexBackfillMerger.
 func NewIndexBackfillMerger(
 	flowCtx *execinfra.FlowCtx, processorID int32, spec execinfrapb.IndexBackfillMergerSpec,

--- a/pkg/sql/execinfra/processorsbase.go
+++ b/pkg/sql/execinfra/processorsbase.go
@@ -57,6 +57,11 @@ type Processor interface {
 	// NB: this method doesn't take the context as parameter because the context
 	// was already captured on Run().
 	Resume(output RowReceiver)
+
+	// Close releases the resources of the processor and possibly its inputs.
+	// Must be called at least once on a given Processor and can be called
+	// multiple times.
+	Close(context.Context)
 }
 
 // DoesNotUseTxn is an interface implemented by some processors to mark that
@@ -733,6 +738,11 @@ func (pb *ProcessorBaseNoHelper) Resume(output RowReceiver) {
 		panic("processor output is not provided for emitting rows")
 	}
 	Run(pb.ctx, pb.self, output)
+}
+
+// Close is part of the Processor interface.
+func (pb *ProcessorBaseNoHelper) Close(context.Context) {
+	pb.self.ConsumerClosed()
 }
 
 // ProcStateOpts contains fields used by the ProcessorBase's family of functions

--- a/pkg/sql/flowinfra/flow.go
+++ b/pkg/sql/flowinfra/flow.go
@@ -672,17 +672,14 @@ func (f *FlowBase) Cleanup(ctx context.Context) {
 
 	// Ensure that all processors are closed. Usually this is done automatically
 	// (when a processor is exhausted or at the end of execinfra.Run loop), but
-	// in edge cases we need to do it here. ConsumerClosed can be called
-	// multiple times.
+	// in edge cases we need to do it here. Close can be called multiple times.
 	//
-	// Note that ConsumerClosed is not thread-safe, but at this point if the
-	// processor wasn't fused and ran in its own goroutine, that goroutine must
-	// have exited since Cleanup is called after having waited for all started
+	// Note that Close is not thread-safe, but at this point if the processor
+	// wasn't fused and ran in its own goroutine, that goroutine must have
+	// exited since Cleanup is called after having waited for all started
 	// goroutines to exit.
 	for _, proc := range f.processors {
-		if rs, ok := proc.(execinfra.RowSource); ok {
-			rs.ConsumerClosed()
-		}
+		proc.Close(ctx)
 	}
 
 	// Release any descriptors accessed by this flow.

--- a/pkg/sql/importer/exportcsv.go
+++ b/pkg/sql/importer/exportcsv.go
@@ -312,6 +312,9 @@ func (sp *csvWriter) Resume(output execinfra.RowReceiver) {
 	panic("not implemented")
 }
 
+// Close is part of the execinfra.Processor interface.
+func (*csvWriter) Close(context.Context) {}
+
 func init() {
 	rowexec.NewCSVWriterProcessor = newCSVWriterProcessor
 }

--- a/pkg/sql/importer/exportparquet.go
+++ b/pkg/sql/importer/exportparquet.go
@@ -260,6 +260,9 @@ func (sp *parquetWriterProcessor) Resume(output execinfra.RowReceiver) {
 	panic("not implemented")
 }
 
+// Close is part of the execinfra.Processor interface.
+func (*parquetWriterProcessor) Close(context.Context) {}
+
 // Resume is part of the execinfra.Processor interface.
 func (sp *parquetWriterProcessor) testingKnobsOrNil() *ExportTestingKnobs {
 	if sp.flowCtx.TestingKnobs().Export == nil {

--- a/pkg/sql/rowexec/columnbackfiller.go
+++ b/pkg/sql/rowexec/columnbackfiller.go
@@ -110,6 +110,9 @@ func (*columnBackfiller) Resume(output execinfra.RowReceiver) {
 	panic("not implemented")
 }
 
+// Close is part of the execinfra.Processor interface.
+func (*columnBackfiller) Close(context.Context) {}
+
 func (cb *columnBackfiller) doRun(ctx context.Context) *execinfrapb.ProducerMetadata {
 	finishedSpans, err := cb.mainLoop(ctx)
 	if err != nil {

--- a/pkg/sql/rowexec/indexbackfiller.go
+++ b/pkg/sql/rowexec/indexbackfiller.go
@@ -438,3 +438,8 @@ func (ib *indexBackfiller) buildIndexEntryBatch(
 func (ib *indexBackfiller) Resume(output execinfra.RowReceiver) {
 	panic("not implemented")
 }
+
+// Close is part of the execinfra.Processor interface.
+func (ib *indexBackfiller) Close(ctx context.Context) {
+	ib.IndexBackfiller.Close(ctx)
+}

--- a/pkg/sql/rowexec/sample_aggregator.go
+++ b/pkg/sql/rowexec/sample_aggregator.go
@@ -213,6 +213,12 @@ func (s *sampleAggregator) Run(ctx context.Context, output execinfra.RowReceiver
 	s.MoveToDraining(nil /* err */)
 }
 
+// Close is part of the execinfra.Processor interface.
+func (s *sampleAggregator) Close(context.Context) {
+	s.input.ConsumerClosed()
+	s.close()
+}
+
 func (s *sampleAggregator) close() {
 	if s.InternalClose() {
 		s.memAcc.Close(s.Ctx())

--- a/pkg/sql/rowexec/sampler.go
+++ b/pkg/sql/rowexec/sampler.go
@@ -488,6 +488,12 @@ func (s *samplerProcessor) sampleRow(
 	return false, nil
 }
 
+// Close is part of the execinfra.Processor interface.
+func (s *samplerProcessor) Close(context.Context) {
+	s.input.ConsumerClosed()
+	s.close()
+}
+
 func (s *samplerProcessor) close() {
 	if s.InternalClose() {
 		s.memAcc.Close(s.Ctx())

